### ISSUE TITLE
[Feature Request]: Labels/Taints for autoscaler_nodepools

### ISSUE THREAD

**Issue by Fabb111** [kube-hetzner/terraform-hcloud-kube-hetzner#726]
### Description

Since autoscaler respects labels and taints for scheduling pods, it would be great to be able to set labels and taints on the autoscaler node pools similar to the normal node pools. If that's already possible somehow, it would be great to document it since I couldn't find anything related to it.

**Comment by mysticaltech** [comment]
Yep, it's possible! Will do ASAP, otherwise PR welcome, it needs to follow the same from as normal agents to propagate the values. So not a lot to be done, but it does need testing!

**Comment by ifeulner** [comment]
We are currently  already applying the default label / taints... maybe a set dedicated to the autoscaling nodes should be considered.
(see code [here](https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/blob/488e107b599d538a732b8d7134b85197c967de38/autoscaler-agents.tf#L82))

**Comment by mysticaltech** [comment]
@ifeulner Yes, definitely, it's actually quite straightforward, just need to spend some time on it.

**Comment by sambonbonne** [comment]
I'm trying to implement this feature but it does not seem that straightforward: the cluster autoscaler can only take one `HCLOUD_CLOUD_INIT` variable, which contains the taints and labels we want to set.

The solution I'm trying to implement is to split `templates/autoscaler.yaml.tpl` to make a dedicated template for the Deployment and template it with different names and labels (`cluster-autoscaler-${autoscaler_pool_name}`) to be able to give each autoscaler a different `HCLOUD_CLOUD_INIT` variable.

It means it will start one autoscaler deployment per autoscaled pool, which may not be desirable to everyone.

Should I continue this way or did I miss something that would make the configuration simpler? If I continue, be assured my pull request will contain some documentation update in order to warn users about this behavior.

**Comment by valkenburg-prevue-ch** [comment]
I don't know of a simpler solution, and I would be impressed and happy if your plan is going to work!

-------- Oorspronkelijk bericht --------
Op 14 jun. 2023 16:36, schreef Sam D :

**Comment by sambonbonne** [comment]
I started working on it today but I currently have an issue: the autoscaler does not trigger the scale up of the first node (even if I set `min_nodes = 1`).

I will continue to work on it next week.

If anyone want to try it, you can change the source of the `kube-hetzner` module in your `kube.tf` to:

```hcl
source = "github.com/sambonbonne/terraform-hcloud-kube-hetzner?ref=feature%2F726-labels-and-taints-for-autoscaler-nodepools"
```

The source if readable on my fork, on a specific branch: [github.com/sambonbonne/terraform-hcloud-kube-hetzner/tree/feature/726-labels-and-taints-for-autoscaler-nodepools](https://github.com/sambonbonne/terraform-hcloud-kube-hetzner/tree/feature/726-labels-and-taints-for-autoscaler-nodepools).

**Comment by ifeulner** [comment]
> I'm trying to implement this feature but it does not seem that straightforward: the cluster autoscaler can only take one `HCLOUD_CLOUD_INIT` variable, which contains the taints and labels we want to set.
> 
> The solution I'm trying to implement is to split `templates/autoscaler.yaml.tpl` to make a loop around the `Deployment` and template it with different names and labels (`cluster-autoscaler-${autoscaler_pool_name}`) to be able to give each autoscaler a different `HCLOUD_CLOUD_INIT` variable.
> 
> It means it will start one autoscaler deployment per autoscaled pool, which may not be desirable to everyone.

Having multiple autoscaler instances per cluster introduces at least more complexity... I think it is even not really supported. 
Seems to be for me too much unknowns for a rare use case.

**Comment by sambonbonne** [comment]
@ifeulner that's what I was thinking when implementing it, I wonder how it will work when someone will want to create multiple autoscaling nodes pools.

I can continue to work on it but I don't know if I will be able to test all the cases, I see at least:

* creating one autoscaling nodes pool without labels and taints
* creating multiple autoscaling nodes pools without labels and taints
* create one autoscaling nodes pool with labels and taints
* creating multiple autoscaling nodes pools with taints and labels

But your comment leads me to the critical question: if I make this contribution, will it be accepted upstream (in your repo)? If not, I'm not sure this is useful for me to continue. I do it because I have a client who needs an autoscaling nodes pool with taints and labels but if the feature is not implemented/accepted in `terraform-hcloud-kube-hetzner`, I won't work on it purposelessly.

**Comment by mysticaltech** [comment]
@sambonbonne Thank you for picking this up. What would definitely be useful at this moment is a way to set custom global taints and labels for all nodepools, and what's needed is just mapping the value from two variables to here:

https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/blob/8b37fbed06e488ec218da14861bc3d68ff5d0a72/autoscaler-agents.tf#L69-L92

So it should be pretty straightforward. As for the per-nodepool ones, it seems hard if not impossible. However, up to you if you can demonstrate that it works properly. However, the above seems like a good middleground.

**Comment by sambonbonne** [comment]
@mysticaltech yes that's what I was thinking, I think I can make it work per-nodepool but it means having one autoscaler `Deployment` per nodepool. @ifeulner seem to not like this solution (and I understand this adds some overhead).

So the main question is: if I implement a solution with one autoscaler per nodepool, may this solution be acceptable for you (by *you*, I mean the maintainer**s** of this repository)?

**Comment by mysticaltech** [comment]
@sambonbonne I just checked, and the Kubernetes autoscaler documentation advises to have a single autoscaler per Kubernetes cluster, especially when using cloud provider implementations. So in that case, it's not a feature we would want to support at this time.

That said, yes we can have globa autoscaler labels/taints that are customizable via variables.

**Comment by sambonbonne** [comment]
@mysticaltech thanks for the clarification. I understand best practices should be respected so I'll implement that (that is not the use case I have but I won't go against the recommendation from the cluster autoscaler developers).

I will create a global variable for autoscaled nodepools labels and taints and make a PR this week.

**Comment by sambonbonne** [comment]
I opened a pull request, I hope I did not forget anything: https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/pull/849

**Comment by mysticaltech** [comment]
Released in v2.2.3! Awesome work @sambonbonne 🙏
